### PR TITLE
Embed vocabularies in to binary

### PR DESCRIPTION
**Pull request by shapor**
This both increases performance and reduces the binary size due to decreased dependencies.

### CODE REVIEW

**Comment by shapor** [comment]
Before:
```ropahs-mba:orig.tiktoken-go shapor$ go test -v
=== RUN   TestRegex2Func
--- PASS: TestRegex2Func (0.00s)
=== RUN   TestEncoding
--- PASS: TestEncoding (0.20s)
=== RUN   TestDecoding
--- PASS: TestDecoding (0.03s)
PASS
ok  	github.com/pkoukk/tiktoken-go	0.355s
```

After:
```
ropahs-mba:tiktoken-go shapor$ go test -v
=== RUN   TestRegex2Func
--- PASS: TestRegex2Func (0.00s)
=== RUN   TestEncoding
--- PASS: TestEncoding (0.03s)
=== RUN   TestDecoding
--- PASS: TestDecoding (0.03s)
PASS
ok  	github.com/shapor/tiktoken-go	0.216s
```

File: load.go [hunk 1]
```diff
@@ -1,105 +1,57 @@
...
-		if line == "" {
-			continue
+var (
+	//go:embed cl100k.gob
+	cl100k []byte
+	//go:embed p50k.gob
+	p50k []byte
+	//go:embed r50k.gob
+	r50k          []byte
+	embedded_maps = func() (s struct {
```
Comment (heatxsink) [inline load.go]: Remove underscores in variable names.

File: load.go [hunk 2]
```diff
@@ -1,105 +1,57 @@
...
+		dec = gob.NewDecoder(bytes.NewReader(p50k))
+		if err := dec.Decode(&s.P50k_base); err != nil {
+			panic(err)
 		}
-		rank, err := strconv.Atoi(parts[1])
-		if err != nil {
-			return nil, err
+		dec = gob.NewDecoder(bytes.NewReader(r50k))
+		if err := dec.Decode(&s.R50k_base); err != nil {
+			panic(err)
```
Comment (heatxsink) [inline load.go]: Why is there a panic here?

File: load.go [hunk 3]
```diff
@@ -1,105 +1,57 @@
...
 		}
-		bpeRanks[string(token)] = rank
-	}
-	return bpeRanks, nil
+		return
+	}()
+)
+
+type BpeLoader interface {
```
Comment (heatxsink) [inline load.go]: Is there a REAL need for an interface this soon?

File: load.go [hunk 4]
```diff
@@ -1,105 +1,57 @@
...
+		if err := dec.Decode(&s.Cl100k_base); err != nil {
+			panic(err)
 		}
-		parts := strings.Split(line, " ")
-		token, err := base64.StdEncoding.DecodeString(parts[0])
-		if err != nil {
-			return nil, err
+		dec = gob.NewDecoder(bytes.NewReader(p50k))
+		if err := dec.Decode(&s.P50k_base); err != nil {
+			panic(err)
```
Comment (heatxsink) [inline load.go]: Why use of panic vs. returning the `err`?

File: load.go (hunk 1 above)
Comment (heatxsink) [reply load.go]: Perhaps it should look like ... CL100Kbase, P50Kbase, R50Kbase, OR embeddedMapsBase and CL100K, P50K, R50K accordingly.

File: load.go [hunk 5]
```diff
@@ -1,105 +1,57 @@
 package tiktoken
```
Comment (heatxsink) [inline load.go]: Consider renaming the package to something without `tik` in it.

**Review by heatxsink — CHANGES_REQUESTED** [review]

**Comment by shapor** [comment]
Updated. PTAL

File: encoding.go [hunk 6]
```diff
@@ -70,6 +83,9 @@ var encodingMap map[string]*Encoding
 var l *sync.Mutex
 
 func init() {
+	gob.NewDecoder(bytes.NewReader(cl100k)).Decode(&vocabCL100K)
```
Comment (heatxsink) [inline encoding.go]: Why not just put this into each function? ex: cl100k_base() ...

**Review by heatxsink — CHANGES_REQUESTED** [review]

File: encoding.go (hunk 6 above)
Comment (shapor) [reply encoding.go]: because it would be inefficient to have to re-decode the gob every time

File: encoding.go [hunk 7]
```diff
@@ -111,9 +124,15 @@ func initEncoding(encodingName string) (*Encoding, error) {
 	}
 }
 
+func decodeVocab(blob []byte, vocab *map[string]int) error {
```
Comment (heatxsink) [inline encoding.go:127]: +1

Comment (heatxsink) [reply encoding.go:127]: I noticed in getEncoding() there is use of a mutex around access to `map[string]int`. This should use `sync.Map` (https://pkg.go.dev/sync#Map). Thoughts?

**Review by heatxsink — CHANGES_REQUESTED** [review]

File: encoding.go (hunk 7 above)
Comment (shapor) [reply encoding.go:127]: From the link:

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

This code is already protected w/ the existing mutex you mentioned.  No need to change anything to maintain safety here, which is exactly what the Map type doc recommends.

**Review by heatxsink — APPROVED** [review]
LGTM

**Pull request merged**